File: app/services/metrics.py
"""
Telemetry — lightweight in-process metrics with rolling percentile window.
Ported (heavily trimmed) from openproxy's analytics: counters per provider,
latency percentiles (p50/p95/p99) over a rolling window, and race stats.

State is kept in memory (per worker) and mirrored to Redis so /api/metrics
can be served by any worker without losing history on restart.
"""

import logging
import time
from collections import deque
from dataclasses import dataclass, field

from app.database.redis import get_redis

logger = logging.getLogger(__name__)

METRICS_KEY = "gw:metrics:summary"
LATENCY_PREFIX = "gw:metrics:latency:"
WINDOW = 512  # samples kept per latency metric


@dataclass
class MetricSample:
    timestamp: float
    latency_ms: float
    tokens: int = 0


@dataclass
class MetricsStore:
    """Rolling-window metrics, one bucket per (provider, model) pair."""

    total_requests: int = 0
    total_errors: int = 0
    total_races: int = 0
    race_wins: int = 0
    per_provider: dict = field(default_factory=dict)
    _latency: dict = field(default_factory=dict)  # key -> deque[MetricSample]
    _started_at: float = field(default_factory=time.time)

    def record(
        self,
        provider: str,
        model: str | None,
        latency_ms: float,
        ok: bool,
        tokens: int = 0,
        raced: bool = False,
        won_race: bool = False,
    ):
        """Record one upstream call."""
        key = f"{provider}:{model or ''}"
        self.total_requests += 1
        if not ok:
            self.total_errors += 1
        if raced:
            self.total_races += 1
            if won_race:
                self.race_wins += 1

        bucket = self.per_provider.setdefault(
            provider,
            {"requests": 0, "errors": 0, "tokens": 0, "models": set()},
        )
        bucket["requests"] += 1
        bucket["errors"] += 0 if ok else 1
        bucket["tokens"] += tokens
        if model:
            bucket["models"].add(model)

        buf = self._latency.setdefault(key, deque(maxlen=WINDOW))
        buf.append(MetricSample(timestamp=time.time(), latency_ms=latency_ms, tokens=tokens))

    def summary(self) -> dict:
        """Aggregate counters + latency percentiles for /api/metrics."""
        def pct(samples, p):
            if not samples:
                return None
            arr = sorted(s.latency_ms for s in samples)
            k = (len(arr) - 1) * (p / 100.0)
            f = int(k)
            c = min(f + 1, len(arr) - 1)
            frac = k - f
            return round(arr[f] + (arr[c] - arr[f]) * frac, 2)

        latency = {}
        for key, buf in self._latency.items():
            samples = list(buf)
            latency[key] = {
                "count": len(samples),
                "p50_ms": pct(samples, 50),
                "p95_ms": pct(samples, 95),
                "p99_ms": pct(samples, 99),
                "avg_ms": round(sum(s.latency_ms for s in samples) / len(samples), 2) if samples else None,
                "min_ms": round(min(s.latency_ms for s in samples), 2) if samples else None,
                "max_ms": round(max(s.latency_ms for s in samples), 2) if samples else None,
            }

        uptime_s = round(time.time() - self._started_at, 1)
        providers = {
            name: {
                "requests": data["requests"],
                "errors": data["errors"],
                "error_rate": round(data["errors"] / data["requests"], 4) if data["requests"] else 0.0,
                "tokens": data["tokens"],
                "models": sorted(data["models"]),
            }
            for name, data in sorted(self.per_provider.items())
        }

        return {
            "uptime_seconds": uptime_s,
            "total_requests": self.total_requests,
            "total_errors": self.total_errors,
            "error_rate": round(self.total_errors / self.total_requests, 4) if self.total_requests else 0.0,
            "total_races": self.total_races,
            "race_wins": self.race_wins,
            "race_win_rate": round(self.race_wins / self.total_races, 4) if self.total_races else 0.0,
            "providers": providers,
            "latency": latency,
        }


metrics = MetricsStore()


async def summarize_persisted() -> dict:
    """Merge the in-memory snapshot with the Redis mirror (multi-worker-safe)."""
    local = metrics.summary()
    try:
        redis = await get_redis()
        raw = await redis.get(METRICS_KEY)
        if not raw:
            return local
        persisted = __import__("json").loads(raw)
        # Merge: keep the larger counters (approximation across restarts)
        merged = dict(local)
        merged["total_requests"] = max(local["total_requests"], persisted.get("total_requests", 0))
        merged["total_errors"] = max(local["total_errors"], persisted.get("total_errors", 0))
        return merged
    except Exception:
        return local


async def persist_snapshot():
    """Best-effort flush of the in-memory summary to Redis (called periodically)."""
    try:
        redis = await get_redis()
        import json
        await redis.set(METRICS_KEY, json.dumps(metrics.summary()))
    except Exception as e:
        logger.debug(f"Metrics persist failed: {e}")

File: gateway/internal/metrics/metrics.go
// Package metrics is the in-process telemetry store.
//
// Latency percentiles are computed over a small rolling window per
// (provider, model) bucket. The global counters sit behind a mutex since
// requests can come from many goroutines.
package metrics

import (
	"sort"
	"sync"
	"time"
)

// Window is how many samples we keep per (provider, model) for percentiles.
const Window = 512

type providerStats struct {
	Requests int
	Errors   int
	Tokens   int
	Models   map[string]struct{}
}

type sample struct {
	at    time.Time
	ms    float64
	tokens int
}

// Store is the global metrics registry.
type Store struct {
	mu           sync.RWMutex
	total        int
	errors       int
	races        int
	raceWins     int
	startedAt    time.Time
	perProvider  map[string]*providerStats
	latency      map[string][]sample // key = provider:model
}

// New returns an empty Store.
func New() *Store {
	return &Store{
		startedAt:   time.Now(),
		perProvider: map[string]*providerStats{},
		latency:     map[string][]sample{},
	}
}

var (
	// global is the singleton — services register to it directly.
	global = New()
)

// Global returns the singleton registry.
func Global() *Store {
	return global
}

// Record registers one upstream call outcome.
func (s *Store) Record(provider, model string, latencyMs float64, ok bool, tokens int, raced, won bool) {
	s.mu.Lock()
	defer s.mu.Unlock()

	s.total++
	if !ok {
		s.errors++
	}
	if raced {
		s.races++
	}
	if won {
		s.raceWins++
	}

	p := s.perProvider[provider]
	if p == nil {
		p = &providerStats{Models: map[string]struct{}{}}
		s.perProvider[provider] = p
	}
	p.Requests++
	if !ok {
		p.Errors++
	}
	p.Tokens += tokens
	if model != "" {
		p.Models[model] = struct{}{}
	}

	key := provider + ":" + model
	buf := s.latency[key]
	if buf == nil {
		buf = make([]sample, 0, Window)
	}
	buf = append(buf, sample{at: time.Now(), ms: latencyMs, tokens: tokens})
	if len(buf) > Window {
		buf = buf[len(buf)-Window:]
	}
	s.latency[key] = buf
}

// Summary renders the registry for /api/metrics.
func (s *Store) Summary() map[string]any {
	s.mu.RLock()
	defer s.mu.RUnlock()

	providers := map[string]any{}
	for name, p := range s.perProvider {
		models := make([]string, 0, len(p.Models))
		for m := range p.Models {
			models = append(models, m)
		}
		sort.Strings(models)
		errRate := 0.0
		if p.Requests > 0 {
			errRate = float64(p.Errors) / float64(p.Requests)
		}
		providers[name] = map[string]any{
			"requests":    p.Requests,
			"errors":      p.Errors,
			"error_rate":  round4(errRate),
			"tokens":      p.Tokens,
			"models":      models,
		}
	}

	latency := map[string]any{}
	for key, buf := range s.latency {
		if len(buf) == 0 {
			continue
		}
		sorted := make([]float64, len(buf))
		for i, s := range buf {
			sorted[i] = s.ms
		}
		sort.Float64s(sorted)
		sum := 0.0
		for _, v := range sorted {
			sum += v
		}
		latency[key] = map[string]any{
			"count":  len(buf),
			"p50_ms": round2(percentile(sorted, 50)),
			"p95_ms": round2(percentile(sorted, 95)),
			"p99_ms": round2(percentile(sorted, 99)),
			"avg_ms": round2(sum / float64(len(buf))),
		}
	}

	errorRate := 0.0
	if s.total > 0 {
		errorRate = float64(s.errors) / float64(s.total)
	}
	winRate := 0.0
	if s.races > 0 {
		winRate = float64(s.raceWins) / float64(s.races)
	}

	return map[string]any{
		"uptime_seconds":   time.Since(s.startedAt).Seconds(),
		"total_requests":   s.total,
		"total_errors":     s.errors,
		"error_rate":       round4(errorRate),
		"total_races":      s.races,
		"race_wins":        s.raceWins,
		"race_win_rate":    round4(winRate),
		"providers":        providers,
		"latency":          latency,
	}
}

func percentile(sorted []float64, p float64) float64 {
	if len(sorted) == 0 {
		return 0
	}
	k := float64(len(sorted)-1) * (p / 100.0)
	lo := int(k)
	hi := lo + 1
	if hi >= len(sorted) {
		hi = len(sorted) - 1
	}
	frac := k - float64(lo)
	return sorted[lo] + (sorted[hi]-sorted[lo])*frac
}

func round2(f float64) float64 { return roundN(f, 2) }
func round4(f float64) float64 { return roundN(f, 4) }

func roundN(f float64, places int) float64 {
	mul := 1.0
	for i := 0; i < places; i++ {
		mul *= 10
	}
	return float64(int(f*mul+0.5)) / mul
}
